swe_roe/ex2.c

```diff
@@ -0,0 +1,1224 @@
+  PetscInt *id;        /* id of the vertex in local numbering                  */
+  PetscInt *global_id; /* global id of the vertex in local numbering */
+
+  PetscBool *is_local; /* true if the vertex is shared by a local cell         */
+
+  PetscInt *num_cells; /* number of cells sharing the vertex          */
+  PetscInt *num_edges; /* number of edges sharing the vertex                   */
+
+  PetscInt *edge_offset; /* offset for edge IDs that share the vertex                       */
+  PetscInt *cell_offset; /* offset for internal cell IDs that share the vertex              */
+
+  PetscInt *edge_ids; /* edge IDs that share the vertex                       */
+  PetscInt *cell_ids; /* internal cell IDs that share the vertex              */
+
+  RDyCoordinate *coordinate; /* (x,y,z) location of the vertex                       */
+} RDyVertex;
+
+typedef struct {
+  PetscInt *id;        /* id of the edge in local numbering         */
+  PetscInt *global_id; /* global id of the edge in local numbering */
+
+  PetscBool *is_local; /* true if the edge : (1) */
+                       /* 1. Is shared by locally owned cells, or   */
+                       /* 2. Is shared by local cell and non-local  */
+                       /*    cell such that global ID of local cell */
+                       /*    is smaller than the global ID of       */
+                       /*    non-local cell */
+
+  PetscInt *num_cells;  /* number of faces that form the edge        */
+  PetscInt *vertex_ids; /* ids of vertices that form the edge        */
+
+  PetscInt *cell_offset; /* offset for ids of cell that share the edge */
+  PetscInt *cell_ids;    /* ids of cells that share the edge          */
+
+  PetscBool *is_internal; /* false if the edge is on the mesh boundary */
+
+  RDyVector     *normal;   /* unit normal vector                        */
+  RDyCoordinate *centroid; /* edge centroid                             */
+
+  PetscReal *length; /* length of the edge                        */
+
+} RDyEdge;
+
+typedef struct RDyMesh {
+  PetscInt dim;
+
+  PetscInt num_cells;
+  PetscInt num_cells_local;
+  PetscInt num_edges;
+  PetscInt num_vertices;
+  PetscInt num_boundary_faces;
+
+  PetscInt max_vertex_cells, max_vertex_faces;
+
+  RDyCell   cells;
+  RDyVertex vertices;
+  RDyEdge   edges;
+
+  PetscInt *closureSize, **closure, maxClosureSize;
+
+  PetscInt *nG2A;  // Mapping of global cells to application/natural cells
+
+} RDyMesh;
+
+struct _n_User {
+  MPI_Comm  comm;
+  char      filename[PETSC_MAX_PATH_LEN];
+  DM        dm;
+  PetscInt  Nt, Nx, Ny;
+  PetscReal dt, dx, dy;
+  PetscReal Lx, Ly;
+  PetscReal hu, hd;
+  PetscReal tiny_h;
+  PetscInt  dof, rank, comm_size;
+  Vec       B, localB;
+  Vec       localX;
+  PetscBool debug, save, add_building;
+  PetscInt  tstep;
+  PetscBool interpolate;
+
+  RDyMesh *mesh;
+};
+
+PetscErrorCode RDyInitialize_IntegerArray_1D(PetscInt *array_1D, PetscInt ndim_1, PetscInt init_value) {
```

Comment:
I bave some thoughts on arrays that we could discuss when we talk about the library itself.